server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -967,10 +961,18 @@ public DataSegment map(int index, ResultSet r, StatementContext ctx) throws SQLE
     // segment mark calls in rapid succession. So the snapshot update is not done outside of database poll at this time.
     // Updates outside of database polls were primarily for the user experience, so users would immediately see the
     // effect of a segment mark call reflected in MetadataResource API calls.
-    dataSourcesSnapshot = DataSourcesSnapshot.fromUsedSegments(
-        Iterables.filter(segments, Objects::nonNull), // Filter corrupted entries (see above in this method).
-        dataSourceProperties
-    );
+
+    ImmutableMap<String, String> dataSourceProperties = createDefaultDataSourceProperties();
+    if (segments == null || segments.isEmpty()) {
+      log.info("No segments found in the database!");
+      dataSourcesSnapshot = DataSourcesSnapshot.fromUsedSegments(Collections.emptyList(), dataSourceProperties);
```

Comment:
I don't know how `segments == null` is possible here, but if `dataSourcesSnapshot` is already non-null, probably this assignment should be skipped. The cause may be intermittent database unavailability or failure. This condition should be logged.

Reply:
I guess it's harmless to put an `if (segments == null)` block that does a `log.wtf` and returns early, will go ahead and modify to do that just in case.